auto-instrumentation/okhttp/okhttp-3.0/library/src/main/java/io/opentelemetry/instrumentation/library/okhttp/v3_0/internal/CachedSupplier.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.library.okhttp.v3_0.internal;
+
+import java.util.function.Supplier;
+
+/**
+ * This class is internal and is hence not for public use. Its APIs are unstable and can change at
+ * any time.
+ */
+public class CachedSupplier<T> implements Supplier<T> {
+    private Supplier<T> supplier;
+    private T instance;
+
+    public static <T> CachedSupplier<T> create(Supplier<T> instance) {
+        return new CachedSupplier<>(instance);
+    }
+
+    private CachedSupplier(Supplier<T> supplier) {
+        this.supplier = supplier;
+    }
+
+    @Override
+    public T get() {
+        synchronized (this) {
+            if (instance == null) {
```

Comment:
Two things stand out with this code:

1. I feel pretty strongly that you shouldn't `synchronzied(this)`. This uses a lock that is potentially exposed to other users of this instance, and because you can't control it, it opens this class up to locking problems (eg. deadlocks). Better to use a dedicated private lock instance.
2. If the supplier returns null, you will null out the supplier and then on next invocation you will get an NPE. Maybe the right move is to just keep the supplier and not null it out...or at least check to see that instance is not-null before doing so.

Reply:
There should also be unit test coverage for a new class like this one.

Reply:
> I feel pretty strongly that you shouldn't synchronzied(this). This uses a lock that is potentially exposed to other users of this instance, and because you can't control it, it opens this class up to locking problems (eg. deadlocks). Better to use a dedicated private lock instance.

Thanks, it's updated now.

> If the supplier returns null, you will null out the supplier and then on next invocation you will get an NPE. Maybe the right move is to just keep the supplier and not null it out...or at least check to see that instance is not-null before doing so.

Good point, though I really think the supplier must never return null, so I've just added a runtime exception in case that happens. I wouldn't like to keep the supplier if it's not needed to avoid wasting memory.

Reply:
> There should also be unit test coverage for a new class like this one.

Thanks, it's added now.